Fix `ActiveModel::Errors#added?`

bogdanvlviv opened the pull request:
`#added?` without options shouldn't try to match by options

See https://github.com/rails/rails/issues/34813#issuecomment-450375840

I know that we are working on https://github.com/rails/rails/pull/32313 that should introduce model for an error object(instead of using Hash), but there is two reasons why we should merge this to master: Improved test that describes `added?` method behavior better and also we should backport it to `5-2-stable` since we made regression in 5.2.2 ad739f59aa0fd81850010cedf992b81b24da9ff1.

Closes #34813

bogdanvlviv committed:
Fix `ActiveModel::Errors#added?`

`#added?` without options shouldn't try to match by options

See https://github.com/rails/rails/issues/34813#issuecomment-450375840

Closes #34813

urkle commented on activemodel/lib/active_model/errors.rb:
```diff
@@ -329,19 +329,23 @@ def add(attribute, message = :invalid, options = {})
 
       if message.is_a? Symbol
-        details[attribute.to_sym].include? normalize_detail(message, options)
+        if options.blank?
```
would something like this be simpler?
```ruby
normalized = normalize_detail(message, options)
details[attribute.to_sym].any? {|e| normalized.all? {|k,v| e[k] == v} } 
```

bogdanvlviv replied:
Well, actually it isn't the same, see example

```ruby
person = Person.new
person.errors.add :name, :too_long, count: 25, actual_count: 42
# should return `false` since we are missing `actual_count: 42`,
# but with your implementation it returns `true`
assert_not person.errors.added? :name, :too_long, count: 2 
```

lulalala commented:
If possible, could this be worked on my branch instead? I've tried to resolve several conflicts in the past year, basically needing to re-understand the intention and re-implement the fix. Each resolve takes many hours to do.

rafaelfranca commented:
Thank you but the current behavior is the expected and the behavior in 5.2.1 was a bug.

rafaelfranca commented:
@bogdanvlviv we could also implement your PR, but that would be a change in the behavior so I don't think we should backport. What do you think? Also a new method like https://github.com/rails/rails/issues/34629#issuecomment-445373052 could make sense.

bogdanvlviv commented:
Sorry, I hadn't noticed https://github.com/rails/rails/issues/34652#issuecomment-445368949 and that https://github.com/rails/rails/commit/ad739f59aa0fd81850010cedf992b81b24da9ff1 isn't regression but a bug fix, so I do agree with you there is nothing to backport.

Aslo, I agree that behaviour now is expected since `added?` should reflect  method call `add`, and our docs clearly expalin that `If the error message requires an option, then it returns +true+ with the correct option, or +false+ with an incorrect or missing option.`:

```ruby
person.errors.add(:name, :greater_than, count: 5)
person.errors.added?(:name, :greater_than) # false
person.errors.added?(:name, :greater_than, count: 5) # true
```

> Also a new method like #34629 (comment) could make sense.

Yeah, it is a good idea to add `def of_kind?(attribute, message = :invalid)`. I'll do in separate PR.

Thanks!
